pkg/virt-api/rest/instancetype_test.go

```diff
@@ -0,0 +1,255 @@
+			}
+
+			vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+				Name: "nonexistent-instancetype",
+			}
+
+			recorder := callExpandInstancetypeApi(vm)
+			_ = ExpectStatusErrorWithCode(recorder, expectedStatusError)
+		})
+
+		It("should fail if VM points to nonexistent preference", func() {
+			instancetypeMethods.FindPreferenceSpecFunc = func(_ *v1.VirtualMachine) (*instancetypev1alpha1.VirtualMachinePreferenceSpec, error) {
+				return nil, fmt.Errorf("preference does not exist")
+			}
+
+			vm.Spec.Preference = &v1.PreferenceMatcher{
+				Name: "nonexistent-preference",
+			}
+
+			recorder := callExpandInstancetypeApi(vm)
+			_ = ExpectStatusErrorWithCode(recorder, expectedStatusError)
+		})
+
+		It("should apply instancetype to VM", func() {
+			instancetypeMethods.FindInstancetypeSpecFunc = func(_ *v1.VirtualMachine) (*instancetypev1alpha1.VirtualMachineInstancetypeSpec, error) {
+				return &instancetypev1alpha1.VirtualMachineInstancetypeSpec{}, nil
+			}
+
+			cpu := &v1.CPU{Cores: 2}
+
+			instancetypeMethods.ApplyToVmiFunc = func(field *k8sfield.Path, instancetypespec *instancetypev1alpha1.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1alpha1.VirtualMachinePreferenceSpec, vmiSpec *v1.VirtualMachineInstanceSpec) instancetype.Conflicts {
+				vmiSpec.Domain.CPU = cpu
+				return nil
+			}
+
+			vm.Spec.Instancetype = &v1.InstancetypeMatcher{
+				Name: "test-instancetype",
+			}
+
+			expectedVm := vm.DeepCopy()
+			expectedVm.Spec.Template.Spec.Domain.CPU = cpu
+
+			recorder := callExpandInstancetypeApi(vm)
+			responseVm := &v1.VirtualMachine{}
+			Expect(json.NewDecoder(recorder.Body).Decode(responseVm)).To(Succeed())
+
+			Expect(responseVm.Spec.Template.Spec).To(Equal(expectedVm.Spec.Template.Spec))
+		})
+
+		It("should apply preference to VM", func() {
+			instancetypeMethods.FindPreferenceSpecFunc = func(_ *v1.VirtualMachine) (*instancetypev1alpha1.VirtualMachinePreferenceSpec, error) {
+				return &instancetypev1alpha1.VirtualMachinePreferenceSpec{}, nil
+			}
+
+			machineType := "test-machine"
+			instancetypeMethods.ApplyToVmiFunc = func(field *k8sfield.Path, instancetypespec *instancetypev1alpha1.VirtualMachineInstancetypeSpec, preferenceSpec *instancetypev1alpha1.VirtualMachinePreferenceSpec, vmiSpec *v1.VirtualMachineInstanceSpec) instancetype.Conflicts {
+				vmiSpec.Domain.Machine = &v1.Machine{Type: machineType}
+				return nil
+			}
+
+			vm.Spec.Preference = &v1.PreferenceMatcher{
+				Name: "test-preference",
+			}
+
+			expectedVm := vm.DeepCopy()
+			expectedVm.Spec.Template.Spec.Domain.Machine = &v1.Machine{Type: machineType}
+
+			recorder := callExpandInstancetypeApi(vm)
+			responseVm := &v1.VirtualMachine{}
+			Expect(json.NewDecoder(recorder.Body).Decode(responseVm)).To(Succeed())
+
+			Expect(responseVm.Spec.Template.Spec).To(Equal(expectedVm.Spec.Template.Spec))
+		})
+
+		It("should fail, if there is a conflict when applying instancetype", func() {
```

Comment:
As above again.

Reply:
As with the removal of an instancetype or preference after a VM is created but before it's started we also don't protect against users updating an instnacetype so it conflicts during an expand.